### ISSUE TITLE
The Frog of CIFAR 10

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7102]
The Frog of CIFAR 10<br>
<br>
https://ift.tt/3ckTsXp<br>
<br>
About 0 Minutes<br>
<br>
<div>
<section>
<center>
<i>Image produced by a CPPN network trained on CIFAR-10's frog class.</i><br>
<i>Same network generates the image at both <code>30x30</code> and <code>1080x1080</code> pixel resolution.</i><br>
<code><br>
<a href="https://github.com/hardmaru/cppn-gan-vae-cifar-tensorflow">GitHub</a></code>
</center>
<h2>Introduction</h2>
<p>In the <a href="https://blog.otoro.net/2016/04/01/generating-large-images-from-latent-vectors/">previous post</a>, we constructed a generative model that can draw MNIST digits at a very high image resolution, using a latent space vector representation of images. The generator network uses the <a href="https://blog.otoro.net/2016/03/25/generating-abstract-patterns-with-tensorflow/">CPPN model</a> to produce images of any specified resolution. Now using the same model from last time, we will try to apply it naively to train on the <a href="https://www.cs.toronto.edu/~kriz/cifar.html">CIFAR-10</a> dataset.</p>
<p>There is still a lot of work to be done, because the current model is nowhere able to fit a large space of possible colour images, but I thought it would be interesting to see what we can generate out of the model as is. Currently, the model tends to draw a few states of images only, so the encoding function of the VAE will not work well, because there is too much variety in the set of training images for this model to handle.</p>
<h3>Exploring different generative architectures</h3>
<p>The architecture of the CPPN limits the space of what types of images are possible to be drawn by the network. Different architectures correspond to different types of images being generated, despite training on similar data sets.</p>
<p>So in addition to modifying the output regression from the MNIST model to have three outputs rather than one to model colour, we also played around with a variety of different architectures for the generator network. Also, using the model as it is to train on the <em>entire</em> CIFAR-10 data would not really work well, because CIFAR-10 contains such a wide range of different images containing different things, from airplanes to boats to birds to frogs. Below is an example of an interesting image coming out of CPPN after training on the entire set with our model.</p>
<center><a width="100%" href="https://blog.otoro.net/assets/20160406/jpeg/examples/abstract_gan.jpeg"><img width="100%" src="https://blog.otoro.net/assets/20160406/jpeg/examples/abstract_gan.jpeg"></a></center>
<p><i>Abstract CIFAR-10?</i></p>
<p>We notice the image is not <a href="https://blog.otoro.net/2016/03/25/generating-abstract-patterns-with-tensorflow/">as random</a> as the images generated from a CPPN network purely with random weights, but still the image doesn&rsquo;t resemble much unless you really read into the abstract interpretations of such an image (you can become an art critic!). What we can try to do is to train the model with just one class within CIFAR-10 and see if we have more luck. Or at least try to get the model to capture some typical structures, such as the colouring or types of shapes we see in images of a certain class.</p>
<p>In our experiments, we play around with different architectures in the <code>generator</code> class in <code>model.py</code>. For example, we can set every layer in the generator to be <code>tanh</code>:</p>
<p><em>Pure tanh generative network</em></p>
<figure>
<table>
<tbody>
<tr>
<td>
<pre>1
2
3
4
5
6</pre>
</td>
<td>
<pre><span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>tanh</span><span>(</span><span>U</span><span>)</span>

<span>for</span> <span>i</span> <span>in</span> <span>range</span><span>(</span><span>1</span><span>,</span> <span>self</span><span>.</span><span>net_depth_g</span><span>):</span>
  <span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>tanh</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>n_network</span><span>,</span> <span>self</span><span>.</span><span>model_name</span><span>+</span><span>'_g_tanh_'</span><span>+</span><span>str</span><span>(</span><span>i</span><span>)))</span>

<span>output</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>sigmoid</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>self</span><span>.</span><span>c_dim</span><span>,</span> <span>self</span><span>.</span><span>model_name</span><span>+</span><span>'_g_'</span><span>+</span><span>str</span><span>(</span><span>self</span><span>.</span><span>net_depth_g</span><span>)))</span><span>
</span></pre>
</td>
</tr>
</tbody>
</table>
</figure>
<p>The picture below was generated from 8 layers of pure <code>tanh</code> layers, and trained on the frog class of CIFAR-10. We can tell as the lines and surface are very curved and smooth. However, this model was prone to instabilities, because as 10 layers of a tanh network wouldn&rsquo;t deal with gradients that well in the GAN training.</p>
<center><a width="100%" href="https://blog.otoro.net/assets/20160406/jpeg/examples/frog.jpeg"><img width="100%" src="https://blog.otoro.net/assets/20160406/jpeg/examples/frog.jpeg"></a></center>
<p><em>Amphibian Origins (2016)</em></p>
<p>To add a bit of structure and rigidity in the generator, I tried to make the first layer a rectifier <code>relu</code> layer, and released this model to train on the truck class of CIFAR-10, to see what sort of image we get.</p>
<p><em><code>relu</code> activation before the <code>tanh</code> generators</em></p>
<figure>
<table>
<tbody>
<tr>
<td>
<pre>1
2
3
4
5
6</pre>
</td>
<td>
<pre><span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>relu</span><span>(</span><span>U</span><span>)</span>

<span>for</span> <span>i</span> <span>in</span> <span>range</span><span>(</span><span>1</span><span>,</span> <span>self</span><span>.</span><span>net_depth_g</span><span>):</span>
  <span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>tanh</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>n_network</span><span>,</span> <span>self</span><span>.</span><span>model_name</span><span>+</span><span>'_g_tanh_'</span><span>+</span><span>str</span><span>(</span><span>i</span><span>)))</span>

<span>output</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>sigmoid</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>self</span><span>.</span><span>c_dim</span><span>,</span> <span>self</span><span>.</span><span>model_name</span><span>+</span><span>'_g_'</span><span>+</span><span>str</span><span>(</span><span>self</span><span>.</span><span>net_depth_g</span><span>)))</span><span>
</span></pre>
</td>
</tr>
</tbody>
</table>
</figure>
<p>Below is a cool sample of the image generated from this type of network. We can see some structure in the image, and the rigid lines are due to the <code>relu</code> layer at the beginning.</p>
<center><a width="100%" href="https://blog.otoro.net/assets/20160406/jpeg/examples/earth.jpeg"><img width="100%" src="https://blog.otoro.net/assets/20160406/jpeg/examples/earth.jpeg"></a></center>
<p><em>Ground Terrain (2016)</em></p>
<p>I like this structural effect from the <code>relu</code> layers, but at the same time I don&rsquo;t want to sacrifice the smoothness offered from the sigmoid type activations, so I played around with the architecture a bit, and decided on a repeating <code>relu</code> and <code>tanh</code> combination to train on the set of frog data in CIFAR-10.</p>
<p><em>Repeating <code>relu</code> and <code>tanh</code> layers</em></p>
<figure>
<table>
<tbody>
<tr>
<td>
<pre>1
2
3
4
5
6
7</pre>
</td>
<td>
<pre><span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>relu</span><span>(</span><span>U</span><span>)</span>

<span>for</span> <span>i</span> <span>in</span> <span>range</span><span>(</span><span>1</span><span>,</span> <span>self</span><span>.</span><span>net_depth_g</span><span>):</span>
  <span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>tanh</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>n_network</span><span>,</span> <span>self</span><span>.</span><span>model_name</span><span>+</span><span>'_g_tanh_'</span><span>+</span><span>str</span><span>(</span><span>i</span><span>)))</span>
  <span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>relu</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>n_network</span><span>,</span> <span>self</span><span>.</span><span>model_name</span><span>+</span><span>'_g_relu_'</span><span>+</span><span>str</span><span>(</span><span>i</span><span>)))</span>

<span>output</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>sigmoid</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>self</span><span>.</span><span>c_dim</span><span>,</span> <span>self</span><span>.</span><span>model_name</span><span>+</span><span>'_g_'</span><span>+</span><span>str</span><span>(</span><span>self</span><span>.</span><span>net_depth_g</span><span>)))</span><span>
</span></pre>
</td>
</tr>
</tbody>
</table>
</figure>
<p>Now let&rsquo;s see how this architecture models a subclass of CIFAR-10.</p>
<h2>CIFAR-10 Frogs</h2>
<p>A note on how we did our training. Images in CIFAR-10 database come as 32x32 pixel images. We subsample a slightly smaller image inside each training image, so that our training image is actually 30x30 pixels. In addition, we randomly flip the images horizontally during training. These tricks can increase the usefulness of the small amount of data available in each class (6000 pictures of frogs in the train + test set).</p>
<p>Samples from CIFAR-10&rsquo;s Frog Class:</p>
<p>After training the modified CPPN-GAN-VAE model on the set of frog data, we can sample a few images from the model to see how they look like:</p>
<p><em>Samples of CPPN Frogs:</em></p>
<p>We see the samples above <em>sort of</em> resembles some amphibious like creature, and there&rsquo;s some shades of colours and patterns that are similar to parts of the sample set, but obviously we have a long way to go. The richness of the colour types and many different styles of frogs have been left out. However, with this set, we can try to upsample the images to 1080x1080 to see what these pixellated samples look in higher resolution.</p>
<p><em>CPPN Frogs at High Resolution:</em></p>
<p>I think at the higher resolution, these images are more interesting than the images generated from the generator with purely random weights. At least it seems to be trying to express some sort of theme or mood to the audience.</p>
<center>
<video width="80%" poster="/assets/20160406/jpeg/frog/frog_30_full.jpeg" height="80%"></video><br>
<i>The space of 'frog' images quite limited in the model.</i>
</center>
<h2>CIFAR-10 Trucks</h2>
<p>I&rsquo;ve also experimented with training the model on the truck class of the CIFAR-10 data. In addition, I have also tried to remove the radius input <span><span><span><span><span>R</span></span></span></span></span> for CPPN, so the only inputs will be the latent vector, and x and y coordinates, <span><span><span><span><span>(</span><span>Z</span><span>,</span><span>X</span><span>,</span><span>Y</span></span></span></span></span>). I found that sometimes the radius constrains the type of image space we can generate, although it gives us the curve effects for free.</p>
<p>Below is a sample of generated images from this model after training on the truck class. The CPPN&rsquo;s space of colours after training on the truck class are a bit different from the colours after training on the frog class. Maybe there are more orange/red/brighter colours in the training data for trucks.</p>
<p><em>CPPN CIFAR Truck Samples at 30x30 and 1080x1080:</em></p>
<p>I&rsquo;ve also attempted to use the VAE encoder to encode original samples into the limited latent space to see what representations the CPPN can generate. Obviously the space is way too limited, but it is interesting to see certain features getting captured by the network.</p>
<p>As we can see, the model is quite limited in the few variety of images it can generate, and it tries to choose the closest state for each sample fed in the autoencoder&rsquo;s input.</p>
<h2>Conclusions</h2>
<p>There is still a lot of work to be done before making this CPPN type of network to be able to generate rich images, compared to simple MNIST digits. I think there is some value to getting them to work to generate rich colour images, because this type of network allows us to upscale to very large resolutions. I&rsquo;ve thought of a few things that might be worthwhile to try in the future:</p>
<ul>
<li>
<p>Modify the VAE encoder network from a fully connected 2 layer softplus network into a deeper CNN that can identify better features from images in a spatial invariant way.</p>
</li>
<li>
<p>Train on a <a href="https://en.wikipedia.org/wiki/Discrete_cosine_transform">discrete cosine transformed</a> version of the images, and have the network generate a DCT representation of new images to be inverted back to spatial domain.</p>
</li>
<li>
<p>Modify generator network to take advantage of residual / stochastic layer type of training. Getting a 10-plus layer network to learn is proving to be very challenging. Any attempt at dropout will also make the network generate very blurry images.</p>
</li>
<li>
<p>Replace fully-connected layers in the generator by sparsely connected layers, or CNN-type connections. Try to have a DCGAN type of architecture but the generator will have to function like a CPPN.</p>
</li>
</ul>
<p>I&rsquo;ve put the code for this demo <a href="https://github.com/hardmaru/cppn-gan-vae-cifar-tensorflow">here</a> on GitHub, in case you want to play around with making CPPNs train on colour images. The code modifies the CPPN-GAN-VAE model previously used for generating high resolution MNIST images to train on the CIFAR-10 set. It&rsquo;s a bit hacked together and I feel a bit shameful for releasing code of this quality, but I hope you will be able to have better success with this than me!</p>
<p>Please let me know if you can get it to do draw interesting things.</p>
</section>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via &#22823;&#12488;&#12525; &#12539; Machine Learning https://blog.otoro.net<br>
<br>
September 22, 2020 at 03:07PM
